### PR TITLE
add LOCAL_SEARCH_PATH configuration (paths in which Ansible will search for local files when source path is relative)

### PR DESCRIPTION
**Pull request by nitschSB**
##### SUMMARY

Add LOCAL_SEARCH_PATH configuration. Colon-separated paths in which Ansible will search for local files when source path is relative.

##### ISSUE TYPE

- Feature Pull Request

##### ADDITIONAL INFORMATION

This gives the user the possibility to add paths for resolving local relative paths.

Example `ANSIBLE_LOCAL_SEARCH_PATH=/my/path/on/local`

```yaml
- copy:
    src: a-file.txt
    dest: /tmp/
```

will now first searched in
 - `/my/path/on/local/files/a-file.txt` and
 - `/my/path/on/local/a-file.txt`

The `ansible_search_path` variable is `['/my/path/on/local', ...]`.

See ansible/ansible-documentation#1101

### CODE REVIEW

File: lib/ansible/config/base.yml [hunk 1]
```diff
@@ -2115,4 +2115,13 @@ VERBOSE_TO_STDERR:
     - section: defaults
       key: verbose_to_stderr
   type: bool
+LOCAL_SEARCH_PATH:
+  version_added: '2.17'
+  name: Additional prefered search paths for local files
+  default: '{{ "" }}'
```
Comment (bcoca) [inline lib/ansible/config/base.yml]: not sure what you are trying to do here, but his is not what i would accept as a default, remove and just check 'None' as a sentinel value

File: lib/ansible/playbook/base.py [hunk 2]
```diff
@@ -778,7 +778,7 @@ def get_search_path(self):
         Return the list of paths you should search for files, in order.
         This follows role/playbook dependency chain.
         '''
-        path_stack = []
+        path_stack = [] + C.LOCAL_SEARCH_PATH
```
Comment (bcoca) [inline lib/ansible/playbook/base.py]: this is not the place you should add this as it influences a lot more than the search path for copy. See the 'dwim' functions.

**Review by bcoca — COMMENTED** [review]
I'm not sure we want this feature in general, the current 'search path' is already  complex and hard to debug and affects many things like plugin loading, lookups, action plugins, inventory, vars plugins, documentation, etc.

A few use cases and explaining why this feature is needed and other workarounds are not possible will help the case for inclusion.

That said, this is also missing tests and a changelog.

File: lib/ansible/config/base.yml (hunk 1 above)
Comment (nitschSB) [reply lib/ansible/config/base.yml]: copy & paste form other `pathspec` sorry... now i set it to `[]`

**Comment by nitschSB** [comment]
> I'm not sure we want this feature in general, the current 'search path' is already complex and hard to debug and affects many things like plugin loading, lookups, action plugins, inventory, vars plugins, documentation, etc.
> 
> A few use cases and explaining why this feature is needed and other workarounds are not possible will help the case for inclusion.
> 
> That said, this is also missing tests and a changelog.

Hi @bcoca,
you are right. Thanks for good feedback. My first attempt was not good. Now I found the right place for the new config: `lib/ansible/plugins/lookup/__init__.py`. I also added some tests and a changelog.

My use case:
To me, configuration consists of key-values but also includes files and templates. Key-values can be managed in the inventory, but files and templates, not so easily. I would like to have a similar option for some files and templates as for key-values in the inventory. Currently, I use `with_first_found: "{{ ansible_inventory_sources | reverse | map('regex_replace', '^(.*)/inventory/?$', '\\1/files/' + src ) | list }}"`, which I would like to replace with a global configuration.

**Comment by nitschSB** [comment]
@bcoca is there something I can do to get rid of `waiting_on_contributor` ;)

**Comment by bcoca** [comment]
I'm still not sure this should be a configuration global, adding these files to a role/play structure already seems enough, you even have `inventory_dir` and `inventory_file` variables if you really want to specify a path that follows inventory instead.

**Comment by nitschSB** [comment]
hi @bcoca, I understand your concerns.

I also found a nice explicit way to cover my use case. I wrote simple action plugins `copy_from_inventory.py` (and `template_from_inventory` and `synchronize_from_inventory`):
```
#!/usr/bin/python
# Make coding more python3-ish, this is required for contributions to Ansible
from __future__ import absolute_import, division, print_function

__metaclass__ = type

import os

from ansible.plugins.action import ActionBase
from ansible.plugins.action.copy import ActionModule as CopyActionModule


class ActionModule(CopyActionModule, ActionBase):
    TRANSFERS_FILES = True
    _inventory_sources = []

    def run(self, tmp=None, task_vars=None):
        self._inventory_sources = task_vars.get("ansible_inventory_sources")
        return super(ActionModule, self).run(tmp, task_vars)

    # override function from ActionBase
    def _find_needle(self, dirname, needle):

        path_stack = []
        for inventory in reversed(self._inventory_sources):
            path_stack.append(os.path.join(inventory, ".."))

        # raise Exception(path_stack) # for testing

        # if missing it will return a file not found exception
        return self._loader.path_dwim_relative_stack(path_stack, dirname, needle)
```

Let me know how I can give the plugins to the ansible community.

Thank you

**Comment by bcoca** [comment]
You don't need my permission to do that, either publish a collection or find pre-existing one to contribute to.

In any case I consider that this ticket is not needed anymore and is fine to close.

**Pull request closed without merging**